File: Java/src/main/java/com/ilinkai/weixin/messaging/MessageMonitorService.java
package com.ilinkai.weixin.messaging;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ilinkai.weixin.*;
import com.ilinkai.weixin.auth.AccountStore;
import com.ilinkai.weixin.auth.SessionGuard;
import com.ilinkai.weixin.models.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/** 消息监控服务 - 长轮询获取微信消息 */
public class MessageMonitorService {
    private static final Logger logger = LoggerFactory.getLogger(MessageMonitorService.class);
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private static final int MAX_CONSECUTIVE_FAILURES = 3;
    private static final int BACKOFF_DELAY_MS = 30_000;
    private static final int RETRY_DELAY_MS = 2_000;

    private final MessageMonitorOptions options;
    private final WeixinApiClient apiClient;
    private final SessionGuard sessionGuard;
    private final MessageProcessor messageProcessor;
    private final String syncBufFilePath;
    private volatile boolean running;

    private Consumer<MessageContext> onMessageReceived;
    private Consumer<Exception> onError;

    public MessageMonitorService(MessageMonitorOptions options) {
        this.options = options;
        WeixinApiClientOptions clientOpts = new WeixinApiClientOptions();
        clientOpts.setBaseUrl(options.baseUrl);
        clientOpts.setCdnBaseUrl(options.cdnBaseUrl);
        clientOpts.setToken(options.token);
        clientOpts.setLongPollTimeoutMs(options.longPollTimeoutMs);
        this.apiClient = new WeixinApiClient(clientOpts);
        this.sessionGuard = new SessionGuard();
        this.messageProcessor = new MessageProcessor(options.cdnBaseUrl);

        String stateDir = getStateDir();
        new File(stateDir).mkdirs();
        this.syncBufFilePath = stateDir + File.separator + AccountStore.normalizeAccountId(options.accountId) + "_sync_buf.json";
    }

    public void setOnMessageReceived(Consumer<MessageContext> handler) { this.onMessageReceived = handler; }
    public void setOnError(Consumer<Exception> handler) { this.onError = handler; }

    private static String getStateDir() {
        String envPath = System.getenv("ILINKAI_STATE_DIR");
        if (envPath != null && !envPath.trim().isEmpty()) return envPath;
        return System.getProperty("user.home") + File.separator + ".ilinkai" + File.separator + "sync";
    }

    private String loadSyncBuf() {
        try {
            if (!new File(syncBufFilePath).exists()) return "";
            String raw = new String(Files.readAllBytes(Paths.get(syncBufFilePath)));
            SyncBufData data = MAPPER.readValue(raw, SyncBufData.class);
            return data != null && data.getUpdatesBuf != null ? data.getUpdatesBuf : "";
        } catch (Exception e) { return ""; }
    }

    private void saveSyncBuf(String syncBuf) {
        try {
            SyncBufData data = new SyncBufData();
            data.getUpdatesBuf = syncBuf;
            Files.write(Paths.get(syncBufFilePath), MAPPER.writeValueAsBytes(data));
        } catch (Exception e) { logger.error("Failed to save sync buf", e); }
    }

    public void start() throws InterruptedException {
        running = true;
        logger.info("Monitor started: baseUrl={} accountId={}", options.baseUrl, options.accountId);

        String syncBuf = loadSyncBuf();
        int consecutiveFailures = 0;

        while (running) {
            try {
                sessionGuard.assertSessionActive(options.accountId);
                ApiModels.GetUpdatesResponse response = apiClient.getUpdates(syncBuf);

                boolean isApiError = (response.getRet() != null && response.getRet() != 0)
                        || (response.getErrCode() != null && response.getErrCode() != 0);

                if (isApiError) {
                    boolean isSessionExpired = (response.getErrCode() != null && response.getErrCode() == SessionGuard.SESSION_EXPIRED_ERROR_CODE)
                            || (response.getRet() != null && response.getRet() == SessionGuard.SESSION_EXPIRED_ERROR_CODE);
                    if (isSessionExpired) {
                        sessionGuard.pauseSession(options.accountId);
                        long pauseMs = sessionGuard.getRemainingPauseMs(options.accountId);
                        logger.error("getUpdates: session expired, pausing for {} min", Math.ceil(pauseMs / 60000.0));
                        if (onError != null) onError.accept(new WeixinSessionPausedException(options.accountId, (int) Math.ceil(pauseMs / 60000.0)));
                        consecutiveFailures = 0;
                        Thread.sleep(pauseMs);
                        continue;
                    }
                    consecutiveFailures++;
                    if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) {
                        consecutiveFailures = 0;
                        Thread.sleep(BACKOFF_DELAY_MS);
                    } else {
                        Thread.sleep(RETRY_DELAY_MS);
                    }
                    continue;
                }

                consecutiveFailures = 0;
                if (response.getGetUpdatesBuf() != null && !response.getGetUpdatesBuf().isEmpty()) {
                    saveSyncBuf(response.getGetUpdatesBuf());
                    syncBuf = response.getGetUpdatesBuf();
                }

                List<WeixinMessage> messages = response.getMessages() != null ? response.getMessages() : new ArrayList<>();
                for (WeixinMessage msg : messages) {
                    MessageContext context = messageProcessor.convertToContext(msg, options.accountId);
                    if (options.mediaDir != null && !options.mediaDir.isEmpty()) {
                        messageProcessor.processMedia(context, options.mediaDir);
                    }
                    if (onMessageReceived != null) onMessageReceived.accept(context);
                }
            } catch (WeixinSessionPausedException e) {
                throw new RuntimeException(e);
            } catch (InterruptedException e) {
                break;
            } catch (Exception err) {
                consecutiveFailures++;
                logger.error("getUpdates error ({}/{})", consecutiveFailures, MAX_CONSECUTIVE_FAILURES, err);
                if (onError != null) onError.accept(err);
                if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) {
                    consecutiveFailures = 0;
                    Thread.sleep(BACKOFF_DELAY_MS);
                } else {
                    Thread.sleep(RETRY_DELAY_MS);
                }
            }
        }
        logger.info("Monitor ended");
    }

    public void stop() { running = false; }

    private static class SyncBufData {
        @JsonProperty("get_updates_buf") public String getUpdatesBuf;
    }

    public static class MessageMonitorOptions {
        public String baseUrl = WeixinApiClient.DEFAULT_BASE_URL;
        public String cdnBaseUrl = WeixinApiClient.DEFAULT_CDN_BASE_URL;
        public String token;
        public String accountId = "";
        public int longPollTimeoutMs = 35_000;
        public String mediaDir;
    }
}

File: Python/ilinkai_weixin/messaging/message_monitor_service.py
"""消息监控服务"""
import json
import logging
import math
import os
import time
from typing import Optional, Callable

from ..weixin_api_client import WeixinApiClient, WeixinApiClientOptions
from ..auth.session_guard import SessionGuard, WeixinSessionPausedException, SESSION_EXPIRED_ERROR_CODE
from ..auth.account_store import AccountStore
from .message_processor import MessageProcessor, MessageContext

logger = logging.getLogger(__name__)

_MAX_CONSECUTIVE_FAILURES = 3
_BACKOFF_DELAY_S = 30
_RETRY_DELAY_S = 2


class MessageMonitorService:
    """消息监控服务 - 长轮询获取微信消息"""

    def __init__(self, base_url: str, cdn_base_url: str, token: str,
                 account_id: str, long_poll_timeout_ms: int = 35000,
                 media_dir: Optional[str] = None):
        self._base_url = base_url
        self._cdn_base_url = cdn_base_url
        self._account_id = account_id
        self._media_dir = media_dir
        self._api_client = WeixinApiClient(WeixinApiClientOptions(
            base_url=base_url, cdn_base_url=cdn_base_url,
            token=token, long_poll_timeout_ms=long_poll_timeout_ms,
        ))
        self._session_guard = SessionGuard()
        self._message_processor = MessageProcessor(cdn_base_url)
        self._running = False

        state_dir = self._get_state_dir()
        os.makedirs(state_dir, exist_ok=True)
        normalized = AccountStore.normalize_account_id(account_id)
        self._sync_buf_file = os.path.join(state_dir, f"{normalized}_sync_buf.json")

        self.on_message_received: Optional[Callable[[MessageContext], None]] = None
        self.on_error: Optional[Callable[[Exception], None]] = None

    @staticmethod
    def _get_state_dir() -> str:
        env_path = os.environ.get('ILINKAI_STATE_DIR')
        if env_path:
            return env_path
        return os.path.join(os.path.expanduser('~'), '.ilinkai', 'sync')

    def _load_sync_buf(self) -> str:
        try:
            if not os.path.exists(self._sync_buf_file):
                return ""
            with open(self._sync_buf_file, 'r') as f:
                data = json.load(f)
            return data.get('get_updates_buf', '')
        except Exception:
            return ""

    def _save_sync_buf(self, sync_buf: str):
        try:
            with open(self._sync_buf_file, 'w') as f:
                json.dump({'get_updates_buf': sync_buf}, f)
        except Exception as e:
            logger.error("Failed to save sync buf: %s", e)

    def start(self):
        """启动监控（阻塞）"""
        self._running = True
        logger.info("Monitor started: baseUrl=%s accountId=%s", self._base_url, self._account_id)

        sync_buf = self._load_sync_buf()
        consecutive_failures = 0

        while self._running:
            try:
                self._session_guard.assert_session_active(self._account_id)
                response = self._api_client.get_updates(sync_buf)

                is_api_error = (response.ret is not None and response.ret != 0) or \
                               (response.errcode is not None and response.errcode != 0)

                if is_api_error:
                    is_session_expired = (response.errcode == SESSION_EXPIRED_ERROR_CODE) or \
                                         (response.ret == SESSION_EXPIRED_ERROR_CODE)
                    if is_session_expired:
                        self._session_guard.pause_session(self._account_id)
                        pause_ms = self._session_guard.get_remaining_pause_ms(self._account_id)
                        logger.error("getUpdates: session expired, pausing for %d min",
                                     math.ceil(pause_ms / 60000))
                        if self.on_error:
                            self.on_error(WeixinSessionPausedException(
                                self._account_id, int(math.ceil(pause_ms / 60000))))
                        consecutive_failures = 0
                        time.sleep(pause_ms / 1000)
                        continue

                    consecutive_failures += 1
                    if consecutive_failures >= _MAX_CONSECUTIVE_FAILURES:
                        consecutive_failures = 0
                        time.sleep(_BACKOFF_DELAY_S)
                    else:
                        time.sleep(_RETRY_DELAY_S)
                    continue

                consecutive_failures = 0
                if response.get_updates_buf:
                    self._save_sync_buf(response.get_updates_buf)
                    sync_buf = response.get_updates_buf

                messages = response.messages or []
                for msg in messages:
                    context = self._message_processor.convert_to_context(msg, self._account_id)
                    if self._media_dir:
                        self._message_processor.process_media(context, self._media_dir)
                    if self.on_message_received:
                        self.on_message_received(context)

            except WeixinSessionPausedException:
                raise
            except KeyboardInterrupt:
                break
            except Exception as err:
                consecutive_failures += 1
                logger.error("getUpdates error (%d/%d): %s",
                             consecutive_failures, _MAX_CONSECUTIVE_FAILURES, err)
                if self.on_error:
                    self.on_error(err)
                if consecutive_failures >= _MAX_CONSECUTIVE_FAILURES:
                    consecutive_failures = 0
                    time.sleep(_BACKOFF_DELAY_S)
                else:
                    time.sleep(_RETRY_DELAY_S)

        logger.info("Monitor ended")

    def stop(self):
        self._running = False
